### PR TITLE
XML to JSON Run ECS Task lambda

### PR DESCRIPTION
**Pull request by kenoir**
### What is this PR trying to achieve?

Depends on: https://github.com/wellcometrust/platform/pull/911

Runs the XML to JSON ECS task from an S3 event stream when a new XML file is uploaded.

### Who is this change for?

💻 

### Have the following been considered/are they needed?

- [ ] Run `terraform apply`.

### CODE REVIEW

File: api/src/main/scala/uk/ac/wellcome/platform/api/finatra/exceptions/CaseClassMappingExceptionMapper.scala [hunk 1]
```diff
@@ -24,13 +24,15 @@ class CaseClassMappingExceptionWrapper @Inject()(
 
   val contextUri: String = s"${apiScheme}://${apiHost}${apiContext}"
 
-  override def toResponse(request: Request, e: CaseClassMappingException): Response = {
+  override def toResponse(request: Request,
```
Comment (kenoir) [inline api/src/main/scala/uk/ac/wellcome/platform/api/finatra/exceptions/CaseClassMappingExceptionMapper.scala]: `scalafmt`!

Comment (alexwlchan) [reply api/src/main/scala/uk/ac/wellcome/platform/api/finatra/exceptions/CaseClassMappingExceptionMapper.scala]: Fixing it!

Comment (alexwlchan) [reply api/src/main/scala/uk/ac/wellcome/platform/api/finatra/exceptions/CaseClassMappingExceptionMapper.scala]: (And working out why it didn't run on my PR)

File: docker/python3.6_ci/build_lambda.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
+
+set -o errexit
+set -o nounset
+
+if [ -d "src" ]
```
Comment (alexwlchan) [inline docker/python3.6_ci/build_lambda.sh]: nit: I’d pull the `lambda_dir` variable out of the `if` block, then reference it here:

```bash
if [[ -d "$lambda_dir" ]]
```

File: docker/python3.6_ci/Dockerfile [hunk 3]
```diff
@@ -12,6 +12,7 @@ RUN pip install moto==1.0.1
 
 COPY is_up_to_date_with_master.py /app/is_up_to_date_with_master.py
 
+COPY build_lambda.sh /app/build_lambda.sh
```
Comment (alexwlchan) [inline docker/python3.6_ci/Dockerfile]: thought: does `COPY . /app` work?

Might save us having to edit the Dockerfile when we add new scripts.

File: miro_preprocessor/Makefile [hunk 4]
```diff
@@ -26,13 +26,25 @@ xml_to_json_converter-publish: xml_to_json_converter-build $(ROOT)/.docker/publi
 		--project=xml_to_json_converter \
 		--infra-bucket=platform-infra
 
-miro_preprocessor-test: xml_to_json_converter-test
+xml_to_json_run_task-test: $(ROOT)/.docker/python3.6_ci
+	docker run \
+		-v $(ROOT)/miro_preprocessor/xml_to_json_run_task:/data \
+		-e OP=test -e FIND_MATCH_PATHS='/data/src' \
```
Comment (alexwlchan) [inline miro_preprocessor/Makefile]: For consistency with `xml_to_json_converter-test`, do we want these paths to be:

```
-v $(ROOT)/miro_preprocessor/xml_to_json_run_task/src:/data \
-e OP=test -e FIND_MATCH_PATHS='/data'
```

?

Comment (alexwlchan) [reply miro_preprocessor/Makefile]: Also, I wonder if it would be helpful to add a second variable to our Makefiles:

```make
ROOT = $(shell git rev-parse --show-toplevel)
PROJECT = $(ROOT)/miro_preprocessor
```

File: miro_preprocessor/iam_policy_document.tf [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  statement {
+    actions = [
+      "s3:Put*",
+    ]
+
+    resources = [
+      "${data.terraform_remote_state.platform.bucket_miro_data_arn}/json",
+    ]
+  }
+}
```
Comment (alexwlchan) [inline miro_preprocessor/iam_policy_document.tf]: 👌 

Nice separation of privileges.

File: miro_preprocessor/xml_to_json_run_task/main.tf [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+module "trigger_application_restart_on_config_change" {
+  source = "../../terraform/lambda/trigger_s3"
+
+  lambda_function_name = "${module.lambda_xml_to_json_run_task.function_name}"
+  lambda_function_arn  = "${module.lambda_xml_to_json_run_task.arn}"
+  s3_bucket_arn        = "${var.bucket_miro_data_arn}"
+  s3_bucket_id         = "${var.bucket_miro_data_id}"
+  filter_prefix        = "xml"
```
Comment (alexwlchan) [inline miro_preprocessor/xml_to_json_run_task/main.tf]: So this has a filter prefix of `xml`, but the role policy has read permissions for a prefix of `source`.

Are they intentionally different? Could we get a brief explanation in the README of the different folders, maybe?

File: miro_preprocessor/xml_to_json_run_task/main.tf [hunk 7]
```diff
@@ -0,0 +1,21 @@
+# Lambda for publishing ECS service schedules to an SNS topic
+
+module "lambda_xml_to_json_run_task" {
+  source     = "../../terraform/lambda"
+  source_dir = "${path.module}/target"
+
+  name        = "xml_to_json_run_task"
+  description = "Run the Miro XML to JSON ECS task"
```
Comment (alexwlchan) [inline miro_preprocessor/xml_to_json_run_task/main.tf]: nit: The more I read it, the more this description bugs me. Maybe "*Start* the Miro XML to JSON ECS task"? Just to make it clear this Lambda isn’t actually doing the work.

File: miro_preprocessor/xml_to_json_run_task/src/conftest.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
+import boto3
```
Comment (alexwlchan) [inline miro_preprocessor/xml_to_json_run_task/src/conftest.py]: Not to be addressed right now, but we should really find a way to stop just copying these files around.

File: miro_preprocessor/xml_to_json_run_task/src/xml_to_json_run_task.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+def command_for_xml_to_json_task(event, cluster_name, container_name, task_definition):
+    bucket_name = event["bucket_name"]
+    object_src = event["object_key"]
+    object_dst = f'{object_src}.txt'
+
+    cmd = [
+        f'--bucket={bucket_name}',
+        f'--src={object_src}',
+        f'--src={object_dst}'
```
Comment (alexwlchan) [inline miro_preprocessor/xml_to_json_run_task/src/xml_to_json_run_task.py]: Should this be `--dst`?

File: miro_preprocessor/xml_to_json_run_task/src/xml_to_json_run_task.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        sns_client,
+        topic_arn,
+        cmd
+    )
+
+
+def main(event, _):
+    print(f'event = {event!r}')
+
+    topic_arn = os.environ["TOPIC_ARN"]
```
Comment (alexwlchan) [inline miro_preprocessor/xml_to_json_run_task/src/xml_to_json_run_task.py]: How do these environment variables end up in the Lambda? I don’t see any associated Terraform config, or have I missed it?

File: miro_preprocessor/xml_to_json_run_task/src/xml_to_json_run_task.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
+# -*- encoding: utf-8 -*-
+
```
Comment (alexwlchan) [inline miro_preprocessor/xml_to_json_run_task/src/xml_to_json_run_task.py]: nit: lose this newline

File: miro_preprocessor/xml_to_json_run_task/src/xml_to_json_run_task.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+"""
+Lambda to run the Miro XML to JSON task when receiving an S3 event.
```
Comment (alexwlchan) [inline miro_preprocessor/xml_to_json_run_task/src/xml_to_json_run_task.py]: minor: I’d rephrase this to make it clear "run the task" means a specific AWS API:

> Lambda to call the RunTask API for the Miro XML-to-JSON ECS task when a file in S3 is updated.

File: miro_preprocessor/xml_to_json_run_task/src/xml_to_json_run_task.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+    s3_events = s3_utils.parse_s3_record(event=event)
+    task_commands = [command_for_xml_to_json_task(
+        event,
+        cluster_name,
+        container_name,
+        task_definition
+    ) for event in s3_events]
+
+    [post_to_sns(sns_client, topic_arn, cmd) for cmd in task_commands]
```
Comment (alexwlchan) [inline miro_preprocessor/xml_to_json_run_task/src/xml_to_json_run_task.py]: Using list comprehensions for side effects in this way is unusual. You’re collecting the results even though you don’t plan to use it – here, a `for` loop is more common, i.e.,

```python
for cmd in task_commands:
    post_to_sns(sns_client, topic_arn, cmd)
```

which makes it explicit that we’re discarding the results.

Comment (alexwlchan) [reply miro_preprocessor/xml_to_json_run_task/src/xml_to_json_run_task.py]: And at this point, I’d drop the `post_to_sns` wrapper, and call `sns_utils.publish_message()` directly. Cut out an unnecessary redirection.

File: miro_preprocessor/xml_to_json_run_task/src/conftest.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+def moto_start(set_region):
+    mock_autoscaling().start()
+    mock_ec2().start()
+    mock_ecs().start()
+    mock_sns().start()
+    mock_sqs().start()
+    yield
+    mock_autoscaling().stop()
+    mock_ec2().stop()
+    mock_ecs().stop()
```
Comment (alicefuzier) [inline miro_preprocessor/xml_to_json_run_task/src/conftest.py]: don't think you need ```
mock_autoscaling()
mock_ec2()
mock_ecs()
```

File: miro_preprocessor/terraform.tf [hunk 15]
```diff
@@ -18,3 +18,13 @@ data "terraform_remote_state" "platform" {
     region = "eu-west-1"
   }
 }
+
+data "terraform_remote_state" "lambda" {
```
Comment (alicefuzier) [inline miro_preprocessor/terraform.tf]: This is only here beacuse of the `lambda_error_alarm_arn` output. I wonder whether this should be in the shared infra instead

File: miro_preprocessor/Makefile (hunk 4 above)
Comment (kenoir) [reply miro_preprocessor/Makefile]: Wouldn't this cause an issue if multiple projects used this variable?

Comment (alexwlchan) [reply miro_preprocessor/Makefile]: Mmm. I think it would still work – `include` just runs the Makefile straight through, so it would only last until the variable was next defined. But might get messy, so let’s leave it for now.

File: miro_preprocessor/terraform.tf (hunk 15 above)
Comment (kenoir) [reply miro_preprocessor/terraform.tf]: 👍  I think we can consolidate this after.

File: miro_preprocessor/xml_to_json_run_task/src/xml_to_json_run_task.py (hunk 10 above)
Comment (kenoir) [reply miro_preprocessor/xml_to_json_run_task/src/xml_to_json_run_task.py]: Yeah - good point, have added these.

File: docker/python3.6_ci/Dockerfile [hunk 16]
```diff
@@ -2,18 +2,9 @@ FROM python:3.6
...
-
-RUN pip install pytest
-
-RUN pip install moto==1.0.1
-
-COPY is_up_to_date_with_master.py /app/is_up_to_date_with_master.py
-
-COPY run.sh /app/run.sh
-COPY test.sh /app/test.sh
+RUN pip install -r /app/requirements.txt
```
Comment (alexwlchan) [inline docker/python3.6_ci/Dockerfile]: 👍 much nicer

File: docker/python3.6_ci/src/build_lambda.sh [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+
+  if [ -e src/requirements.txt ]
+  then
+    echo "Found requirements.txt, installing."
+    pip install -r src/requirements.txt --target=$target_dir
+  else
+    echo "No requirements.txt present. Skipping."
+  fi
+else
+  echo "No ./src directory found! Exiting."
```
Comment (alexwlchan) [inline docker/python3.6_ci/src/build_lambda.sh]: nit: let’s use `src` throughout

File: docker/python3.6_ci/src/run.sh [hunk 18]
```diff
@@ -27,6 +35,7 @@ function run_tests {
   echo "Testing ..."
 
   install_dependencies
+  FIND_MATCH_PATHS=${FIND_MATCH_PATHS:=/data}
```
Comment (alexwlchan) [inline docker/python3.6_ci/src/run.sh]: See here: https://stackoverflow.com/a/27445490/1558022

We want to assign this even if the variable is unset, not just set and null.

File: miro_preprocessor/README.md [hunk 19]
```diff
@@ -13,4 +13,10 @@ An ECS Task with inputs:
 - S3 Source location (a Miro formatted XML dump)
 - S3 Target location
 
-This task downloads and converts a Miro formatted XML dump to JSON and puts it at the S3 target location specified.
\ No newline at end of file
+This task downloads and converts a Miro formatted XML dump to JSON and puts it at the S3 target location specified.
+
+## xml_to_json_run_task
+
+A lambda which is triggered by an S3 upload of a new Miro xml dump.
```
Comment (alexwlchan) [inline miro_preprocessor/README.md]: Detail on where it should be uploaded to?

File: terraform/ecs_script_task/templates/default.json.template [hunk 20]
```diff
@@ -4,7 +4,7 @@
     "essential": true,
     "image": "${app_uri}",
     "memory": ${memory},
-    "name": "app",
+    "name": "${name}",
```
Comment (alexwlchan) [inline terraform/ecs_script_task/templates/default.json.template]: Why has this become a variable?

File: miro_preprocessor/main.tf [hunk 21]
```diff
@@ -1,56 +1,24 @@
...
-    actions = [
-      "s3:Put*",
-    ]
+  s3_read_miro_data_json = "${data.aws_iam_policy_document.s3_read_miro_data.json}"
 
-    resources = [
-      "${data.terraform_remote_state.platform.bucket_miro_data_arn}/json",
-    ]
-  }
+  task_defintion_arn = "${module.xml_to_json_converter.task_arn}"
```
Comment (alexwlchan) [inline miro_preprocessor/main.tf]: `task_defintion_arn`?

`definition`?

File: miro_preprocessor/xml_to_json_run_task/main.tf [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+  description = "Start the Miro XML to JSON ECS task"
+
+  alarm_topic_arn = "${var.lambda_error_alarm_arn}"
+
+  environment_variables = {
+    TOPIC_ARN           = "${var.topic_arn}"
+    CLUSTER_NAME        = "${var.cluster_name}"
+    CONTAINER_NAME      = "${var.container_name}"
+    TASK_DEFINITION_ARN = "${var.task_defintion_arn}"
+  }
```
Comment (alexwlchan) [inline miro_preprocessor/xml_to_json_run_task/main.tf]: 👍

File: terraform/ecs_script_task/templates/default.json.template (hunk 20 above)
Comment (kenoir) [reply terraform/ecs_script_task/templates/default.json.template]: We need to pass this along to the run task lambda and i prefer this approach to hardcoding the value.

File: miro_preprocessor/main.tf (hunk 21 above)
Comment (kenoir) [reply miro_preprocessor/main.tf]: Good point - task_arn actually has a different meaning in AWS as the `arn` of a running instance of a task definition, as distinct from the definition itself.

Comment (kenoir) [reply miro_preprocessor/main.tf]: In case it's not clear - this should be `task_defintion_arn`.
